_posts/2025-05-13-simple-search.md

```diff
@@ -0,0 +1,226 @@
+---
+title: "A simple search engine from scratch*"
+layout: post
+---
+
+\*if you include word2vec.
+
+[Chris](https://www.chrisgregory.me/) and I spent a couple hours the other day
+creating a search engine for my blog from "scratch". Mostly he walked me
+through it because I only vaguely knew what word2vec was before this experiment.
+
+The search engine we made is built on *word embeddings*. This refers to some
+function that takes a word and maps it onto N-dimensional space (in this case,
+N=300) where each dimension vaguely corresponds to some axis of meaning.
+
+The idea behind the search engine is to embed each of my posts into this domain
+by adding up the embeddings for the words in the post. For a given
+search, we'll embed the search the same way. Then we can rank all posts by
+their [cosine similarities](https://en.wikipedia.org/wiki/Cosine_similarity)
+to the query.
+
+The equation below might look scary but it's saying that the cosine similarity,
+which is the cosine of the angle between the two vectors `cos(theta)`, is
+defined as the dot product divided by the product of the magnitudes of each
+vector. We'll walk through it all in detail.
+
+<figure>
+<img src="/assets/img/cosine-similarity.svg" />
+<figcaption>Equation from Wikimedia's <a href="https://en.wikipedia.org/wiki/Cosine_similarity">Cosine similarity</a>
+page.</figcaption>
+</figure>
+
+This is just one metric for query/page similarity. Other metrics include
+[tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) or just plain substring
```

Comment:
We can get fancier than simple cosine similarity. Let's imagine that all of our documents talk about computers, but only one of them talks about compilers (wouldn't that be sad). If one of our search terms is "computer" that doesn't really help narrow down the search and is noise in our embeddings. To reduce noise we can employ a technique called TF-IDF (term frequency inverse document frequency) where we factor out common words across documents and pay closer attention to words that are more unique to each document.